sklearn/cluster/_bisect_k_means.py

```diff
@@ -508,18 +508,25 @@ def _predict_recursive(self, X, sample_weight, cluster_node):
             self._n_threads,
             return_inertia=False,
         )
-        mask = cluster_labels == 0
+        first_cluster_mask = cluster_labels == 0
 
         # Compute the labels for each subset of the data points.
         labels = np.full(X.shape[0], -1, dtype=np.int32)
 
-        labels[mask] = self._predict_recursive(
-            X[mask], sample_weight[mask], cluster_node.left
-        )
+        if first_cluster_mask.any():
```

Comment:
Does this PR prune some recursive calls, making the overall potentially shorter?

Reply:
It does prune silently failing recursive calls. But those could fail right at the beginning so I do not expect any significant performance impact.

Reply:
Actually, maybe even a negligible slowdown because `first_cluster_mask.any()` is not free. But that's probably negligible compared to the actually computation.